src/runner.js

```diff
@@ -28,12 +28,96 @@ const debugResponse = _debug('remote-s3:Runner:res');
  * actually transmitted matched a pre-computed size and hash
  */
 class Runner {
-  constructor(agent, agentOpts) {
-    this.agent = agent || new https.Agent(agentOpts || {}); 
+  constructor(opts) {
+    opts = runSchema(opts || {}, Joi.object().keys({
+      agent: Joi.object(),
+      agentOpts: Joi.object(),
+      maxRetries: Joi.number().min(0).max(10).default(5),
+    }).without('agent', 'agentOpts').optionalKeys(['agent', 'agentOpts']));
+
+    this.agent = opts.agent || new https.Agent(opts.agentOpts || {}); 
+
+    this.maxRetries = opts.maxRetries;
+
   }
 
+  /**
+   * Run a request and retry if a retryable error occurs.  Only 500-series
+   * responses are considered retriable, so any other status code will return
+   * with its status.
+   *
+   * http://docs.aws.amazon.com/general/latest/gr/api-retries.html
+   */
   async run(opts) {
-    return this.runOnce(opts);
+    let {req, body, streamingOutput, followRedirect, currentRedirect, maxRedirects} = opts;
+    let current = 0;
+
+    currentRedirect = currentRedirect || 0;
+    maxRedirects = maxRedirects || 10;
+
+    // Since streams aren't seekable in node, we want to ensure that
+    // whenever 
+    if (body && body instanceof stream.Readable || body && typeof body.pipe === 'function') {
+      debug('WARNING: instances of stream.Readable are not retryable or redirectable');
```

Comment:
on second thought, a file descriptor is tied to the current offset in the file.  That means if we had two streams on the same file descriptor, each readstream would be moving the offset when it does its reads.

For things which *could* be seeked with e.g. a file on the local disk, I think the closure option really is the best.  For files the closure we'd pass as body is `() => fs.createReadStream(path, {start, end})`.  Basically anything we can `.seek(0)` we can also create a closure to recreate it.

This library should be considered pretty low level.  There will be a small wrapper library of *this* library which implements a taskcluster-queue specific uploader client.  I'm doing it this way to ensure that this library is general purpose, so could be used by others.

We also absolutely have to support streams for this library.  If we don't support streams, we will have to read every artifact into memory in order to send it.  That means a 2.2GB image artifact will use 2.2GB memory + overhead...